ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/PullQueryExecutor.java

```diff
@@ -178,7 +179,11 @@ public TableRowsEntity execute(
     try {
       final RoutingOptions routingOptions = new ConfigRoutingOptions(
           statement.getConfig(), statement.getConfigOverrides(), statement.getRequestProperties());
-      final boolean isAlreadyForwarded = routingOptions.skipForwardRequest();
+      // If internal listeners are in use, we require the request to come from that listener to
+      // treat it as having been forwarded.
+      final boolean isAlreadyForwarded = routingOptions.skipForwardRequest() &&
+          // Trust the forward request option if isInternalRequest isn't available.
+          isInternalRequest.orElse(true);
```

Comment:
In which scenarios would `isInternalRequest` not be available? Don't we want these two variables (`skipForwardRequest` and  `isInternalRequest`) to always either both be true or both be false? Which makes me question whether we need both.

Reply:
isInternalRequest wouldn't be available when the internal listener functionality is not enabled, so we still would want `skipForwardRequest` in that case where it's running in a trusted setting and still wants to know to only forward the request once.  But yes, you're right that they aim to infer the identical thing -- that this request is forwarded rather than from the end user.